File: C#/第二套/11900-1060306/11900-1060306/Services/IdCardValidatorService.cs
namespace IdCardChecker.Services;

using IdCardChecker.Models;
using System.Text.RegularExpressions;

public interface IIdCardValidatorService
{
    IdCardRecord Validate(IdCardRecord record);
}

public class IdCardValidatorService : IIdCardValidatorService
{
    private static readonly Dictionary<char, int> LetterMap = new()
    {
        {'A', 10}, {'B', 11}, {'C', 12}, {'D', 13}, {'E', 14}, {'F', 15}, {'G', 16}, {'H', 17},
        {'J', 18}, {'K', 19}, {'L', 20}, {'M', 21}, {'N', 22}, {'P', 23}, {'Q', 24}, {'R', 25},
        {'S', 26}, {'T', 27}, {'U', 28}, {'V', 29}, {'X', 30}, {'Y', 31}, {'W', 32}, {'Z', 33},
        {'I', 34}, {'O', 35}
    };

    /// <summary>
    /// Validates an ID card record.
    /// </summary>
    /// <param name="record">The record to validate.</param>
    /// <returns>The record with error message if any.</returns>
    public IdCardRecord Validate(IdCardRecord record)
    {
        // (A) FORMAT ERROR
        if (!IsValidFormat(record.Id))
        {
            record.Error = "FORMAT ERROR";
            return record;
        }

        // (B) SEX CODE ERROR
        if (!IsValidSex(record.Id, record.Sex))
        {
            record.Error = "SEX CODE ERROR";
            return record;
        }

        // (C) CHECK SUM ERROR
        if (!IsValidChecksum(record.Id))
        {
            record.Error = "CHECK SUM ERROR";
            return record;
        }

        record.Error = string.Empty;
        return record;
    }

    private static bool IsValidFormat(string id)
    {
        if (id.Length != 10) return false;
        if (!char.IsAsciiLetterUpper(id[0])) return false;
        for (int i = 1; i < 10; i++)
        {
            if (!char.IsDigit(id[i])) return false;
        }
        return true;
    }

    private static bool IsValidSex(string id, string sex)
    {
        char sexDigit = id[1];
        if (sex == "M" && sexDigit == '1') return true;
        if (sex == "F" && sexDigit == '2') return true;
        return false;
    }

    private static bool IsValidChecksum(string id)
    {
        if (!LetterMap.TryGetValue(id[0], out int code)) return false;

        int x1 = code / 10;
        int x2 = code % 10;

        int sum = x1 + 9 * x2;
        int[] weights = { 8, 7, 6, 5, 4, 3, 2, 1 };

        for (int i = 0; i < 8; i++)
        {
            sum += (id[i + 1] - '0') * weights[i];
        }

        sum += (id[9] - '0');

        return sum % 10 == 0;
    }
}

File: VB/第二套/11900-1060306/Services/IdCardValidatorService.vb
Imports System.Collections.Generic

Namespace Services
    Public Interface IIdCardValidatorService
        Function Validate(record As Models.IdCardRecord) As Models.IdCardRecord
    End Interface

    Public Class IdCardValidatorService
        Implements IIdCardValidatorService

        Private Shared ReadOnly LetterMap As New Dictionary(Of Char, Integer) From {
            {"A"c, 10}, {"B"c, 11}, {"C"c, 12}, {"D"c, 13}, {"E"c, 14}, {"F"c, 15}, {"G"c, 16}, {"H"c, 17},
            {"J"c, 18}, {"K"c, 19}, {"L"c, 20}, {"M"c, 21}, {"N"c, 22}, {"P"c, 23}, {"Q"c, 24}, {"R"c, 25},
            {"S"c, 26}, {"T"c, 27}, {"U"c, 28}, {"V"c, 29}, {"X"c, 30}, {"Y"c, 31}, {"W"c, 32}, {"Z"c, 33},
            {"I"c, 34}, {"O"c, 35}
        }

        Public Function Validate(record As Models.IdCardRecord) As Models.IdCardRecord Implements IIdCardValidatorService.Validate
            ' (A) FORMAT ERROR
            If Not IsValidFormat(record.Id) Then
                record.ErrorMessage = "FORMAT ERROR"
                Return record
            End If

            ' (B) SEX CODE ERROR
            If Not IsValidSex(record.Id, record.Sex) Then
                record.ErrorMessage = "SEX CODE ERROR"
                Return record
            End If

            ' (C) CHECK SUM ERROR
            If Not IsValidChecksum(record.Id) Then
                record.ErrorMessage = "CHECK SUM ERROR"
                Return record
            End If

            record.ErrorMessage = String.Empty
            Return record
        End Function

        Private Shared Function IsValidFormat(id As String) As Boolean
            If id.Length <> 10 Then Return False
            If Not Char.IsLetter(id(0)) Then Return False
            
            ' VB doesn't have isAsciiLetterUpper, check A-Z
            Dim firstChar As Char = Char.ToUpper(id(0))
            If firstChar < "A"c OrElse firstChar > "Z"c Then Return False

            For i As Integer = 1 To 9
                If Not Char.IsDigit(id(i)) Then Return False
            Next
            Return True
        End Function

        Private Shared Function IsValidSex(id As String, sex As String) As Boolean
            Dim sexDigit As Char = id(1)
            If sex = "M" AndAlso sexDigit = "1"c Then Return True
            If sex = "F" AndAlso sexDigit = "2"c Then Return True
            Return False
        End Function

        Private Shared Function IsValidChecksum(id As String) As Boolean
            Dim firstChar As Char = Char.ToUpper(id(0))
            If Not LetterMap.ContainsKey(firstChar) Then Return False

            Dim code As Integer = LetterMap(firstChar)
            Dim x1 As Integer = code \ 10
            Dim x2 As Integer = code Mod 10

            Dim sum As Integer = x1 + 9 * x2
            Dim weights As Integer() = {8, 7, 6, 5, 4, 3, 2, 1}

            For i As Integer = 0 To 7
                sum += (Val(id(i + 1))) * weights(i)
            Next

            sum += Val(id(9))

            Return sum Mod 10 = 0
        End Function
    End Class
End Namespace
